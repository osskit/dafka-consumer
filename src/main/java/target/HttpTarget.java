package target;

import configuration.Config;
import configuration.TopicsRoutes;
import dev.failsafe.RetryPolicy;
import dev.failsafe.okhttp.FailsafeCall;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.CompletableFuture;
import kafka.Producer;
import monitoring.Monitor;
import okhttp3.*;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.json.JSONArray;
import org.json.JSONObject;

public class HttpTarget implements ITarget {

    public static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    private final TopicsRoutes topicsRoutes;

    private static final Duration httpTimeout = Duration.ofMillis(Config.TARGET_TIMEOUT_MS);

    private static final OkHttpClient client = new OkHttpClient.Builder()
        .callTimeout(httpTimeout)
        .readTimeout(httpTimeout)
        .writeTimeout(httpTimeout)
        .connectTimeout(httpTimeout)
        .build();

    private final Producer producer;

    public HttpTarget(TopicsRoutes topicsRoutes, Producer producer) {
        this.topicsRoutes = topicsRoutes;
        this.producer = producer;
    }

    List<String> cloudEventHeaders = new ArrayList<>() {
        {
            add("ce_id");
            add("ce_time");
            add("ce_specversion");
            add("ce_type");
            add("ce_source");
        }
    };

    public CompletableFuture<Object> call(final ConsumerRecord<String, String> record) {
        var body = RequestBody.create(record.value(), JSON);

        var requestBuilder = new Request.Builder()
            .url(Config.TARGET_BASE_URL + this.topicsRoutes.getRoute(record.topic()))
            .post(body)
            .header("x-record-topic", record.topic())
            .header("x-record-partition", String.valueOf(record.partition()))
            .header("x-record-offset", String.valueOf(record.offset()))
            .header("x-record-timestamp", String.valueOf(record.timestamp()))
            .header("x-record-original-topic", this.getOriginalTopic(record));

        if (Config.ALLOW_BODY_HEADERS) {
            JSONObject jsonObject = new JSONObject(record.value());

            if (jsonObject.has("headers")) {
                JSONObject headersObject = jsonObject.getJSONObject("headers");

                for (String key : headersObject.keySet()) {
                    String value = headersObject.getString(key);

                    requestBuilder.header(key, value);
                }
            }
        }

        record
            .headers()
            .forEach(header -> {
                String headerKey = header.key();

                if (cloudEventHeaders.contains(headerKey)) {
                    headerKey = headerKey.replace("_", "-");
                }

                requestBuilder.header(headerKey, new String(header.value(), StandardCharsets.UTF_8));
            });

        final var request = requestBuilder.build();

        var call = client.newCall(request);
        var requestId = UUID.randomUUID().toString();

        var delay = Config.RETRY_POLICY_EXPONENTIAL_BACKOFF.get(0);
        var maxDelay = Config.RETRY_POLICY_EXPONENTIAL_BACKOFF.get(1);
        var delayFactor = Config.RETRY_POLICY_EXPONENTIAL_BACKOFF.get(2);
        var maxDuration = Duration.ofMillis(Config.RETRY_POLICY_MAX_DURATION_MS);

        Monitor.processMessageStarted(record, requestId);

        var retryPolicy = RetryPolicy
            .<Response>builder()
            .withBackoff(delay, maxDelay, ChronoUnit.MILLIS, delayFactor)
            .handleIf(e -> true)
            .handleResultIf(r -> Integer.toString(r.code()).matches(Config.RETRY_PROCESS_WHEN_STATUS_CODE_MATCH))
            .withMaxDuration(maxDuration)
            .onRetry(e -> {
                Monitor.targetExecutionRetry(
                    Optional
                        .ofNullable(e.getLastResult())
                        .flatMap(r -> {
                            try {
                                var responseBody = r.body().string();
                                return Optional.of(responseBody);
                            } catch (IOException error) {
                                return Optional.empty();
                            }
                        }),
                    e.getLastException(),
                    e.getAttemptCount(),
                    requestId
                );
            })
            .build();

        final long executionStart = (new Date()).getTime();

        return FailsafeCall
            .with(retryPolicy)
            .compose(call)
            .executeAsync()
            .handleAsync((response, throwable) -> {
                if (throwable != null) {
                    Monitor.processMessageError(throwable, requestId);
                    if (Config.RETRY_TOPIC != null) {
                        Monitor.retryProduced(Config.RETRY_TOPIC, requestId);
                        return producer.produce(Config.RETRY_TOPIC, record, requestId);
                    }
                    return CompletableFuture.failedFuture(throwable);
                }

                var statusCode = Integer.toString(response.code());

                if (statusCode.matches(Config.PRODUCE_TO_RETRY_TOPIC_WHEN_STATUS_CODE_MATCH)) {
                    if (Config.RETRY_TOPIC != null) {
                        Monitor.retryProduced(Config.RETRY_TOPIC, requestId);
                        return producer.produce(Config.RETRY_TOPIC, record, requestId);
                    }
                }

                if (statusCode.matches(Config.PRODUCE_TO_DEAD_LETTER_TOPIC_WHEN_STATUS_CODE_MATCH)) {
                    if (Config.DEAD_LETTER_TOPIC != null) {
                        Monitor.deadLetterProduced(Config.DEAD_LETTER_TOPIC, requestId);
                        return producer.produce(Config.DEAD_LETTER_TOPIC, record, requestId);
                    }
                }

                Monitor.processMessageSuccess(requestId, executionStart);

                return CompletableFuture.completedFuture(null);
            });
    }
}
