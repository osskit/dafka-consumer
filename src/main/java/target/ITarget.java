package target;

import configuration.Config;
import java.net.http.HttpRequest.Builder;
import java.util.Iterator;
import java.util.concurrent.CompletableFuture;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.common.header.Header;
import org.json.JSONObject;

public interface ITarget {
    CompletableFuture<TargetResponse> call(ConsumerRecord<String, String> record);

    default String getOriginalTopic(ConsumerRecord<String, String> record) {
        Iterator<Header> headers = record.headers().headers(Config.ORIGINAL_TOPIC).iterator();
        if (headers.hasNext()) {
            return new String(headers.next().value());
        }
        return record.topic();
    }

    default void copyRecordHeaders(Builder httpBuilder, ConsumerRecord<String, String> record) {
        if (record.headers() != null) {
            Iterator<Header> headers = record.headers().iterator();
            while (headers.hasNext()) {
                Header header = headers.next();
                httpBuilder.setHeader(header.key(), header.value().toString());
            }
        }
    }

    default String getRecordCorrelationId(ConsumerRecord<String, String> record) {
        if (record.headers() != null) {
            Iterator<Header> headers = record.headers().iterator();
            while (headers.hasNext()) {
                Header header = headers.next();
                if (header.key().equals(Config.CORRELATION_ID_HEADER_KEY)) {
                    return new String(header.value());
                }
            }
        }

        return null;
    }
}
