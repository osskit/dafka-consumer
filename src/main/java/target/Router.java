package target;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import org.json.JSONObject;
import reactor.kafka.receiver.ReceiverRecord;

import java.util.ArrayList;
import java.util.List;

public class Router {
    private static List<TopicRoute> topicRoutes;

    public Router(String jsonRouter) {
        topicRoutes = new Gson().fromJson(jsonRouter,new TypeToken<ArrayList<TopicRoute>>(){}.getType());
    }

    public Route getRoute(ReceiverRecord<String, String> record) {
        var recordObj = new JSONObject(record.value());
        var y = topicRoutes.stream()
                .filter(x -> x.topic.equals(record.topic()) && recordObj.get(x.filter.field) == x.filter.value)
                .map(x -> {
                    var foo = new Route();
                    foo.endpoint = x.endpoint;
                    foo.body = new JSONObject(record.value()).getJSONObject(x.projection).toString();
                    return foo;
                }).toList();

        return y.get(0);

    }
}

class Route {
    public String endpoint;
    public String body;
}

class Filter {
    public String field;
    public String value;
}

class TopicRoute {
    public String topic;
    public Filter filter;
    public String projection;
    public String endpoint;
}
