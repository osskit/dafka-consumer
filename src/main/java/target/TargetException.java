package target;

import configuration.Config;
import org.apache.kafka.common.header.Header;
import org.apache.kafka.common.header.Headers;
import org.apache.kafka.common.header.internals.RecordHeaders;
import reactor.kafka.receiver.ReceiverRecord;

public class TargetException extends Exception {

    private final Integer responseCode;
    private final String responseBody;
    private final Throwable error;

    public TargetException(Integer responseCode, String responseBody, Throwable error) {
        this.responseCode = responseCode;
        this.responseBody = responseBody;
        this.error = error;
    }

    public Headers getHeaders(ReceiverRecord<String, String> record) {
        Headers headers = record.headers();
        Header originalTopic = headers.lastHeader(Config.ORIGINAL_TOPIC);
        Headers headersToSend = new RecordHeaders();

        for (Header header : headers) {
            headersToSend.add(header);
        }

        if (originalTopic != null) {
            return headersToSend;
        }

        headersToSend.add(Config.ORIGINAL_TOPIC, record.topic().getBytes());
        headersToSend.add("x-group-id", Config.GROUP_ID.getBytes());
        headersToSend.add("x-response-status-code", String.valueOf(responseCode).getBytes());
        headersToSend.add("x-response-body", responseBody.getBytes());
        if (error != null) {
            headersToSend.add("x-unexpected-exception-message", error.getMessage().getBytes());
        }

        return headersToSend;
    }
}
