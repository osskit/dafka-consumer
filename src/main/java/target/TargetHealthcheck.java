package target;

import configuration.Config;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import monitoring.Monitor;

public class TargetHealthcheck {

    private static final HttpClient client = HttpClient.newHttpClient();

    public String getEndpoint() {
        return Config.TARGET_HEALTHCHECK;
    }

    public boolean check() throws IOException {
        if (Config.TARGET_HEALTHCHECK != null) {
            try {
                final var request = HttpRequest
                    .newBuilder()
                    .version(HttpClient.Version.HTTP_1_1)
                    .GET()
                    .uri(URI.create(Config.TARGET_BASE_URL + Config.TARGET_HEALTHCHECK))
                    .build();

                var targetHealthcheckResponse = client.send(request, HttpResponse.BodyHandlers.ofString());
                if (targetHealthcheckResponse.statusCode() != 200) {
                    Monitor.targetHealthcheckFailed(targetHealthcheckResponse.statusCode());
                    return false;
                }
                Monitor.targetHealthcheckPassed(targetHealthcheckResponse.statusCode());
                return true;
            } catch (Exception e) {
                Monitor.initializationError(e);
                return false;
            }
        }
        return true;
    }
}
