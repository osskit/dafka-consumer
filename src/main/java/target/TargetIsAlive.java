package target;

import configuration.Config;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import monitoring.Monitor;

public class TargetIsAlive {

    private static final HttpClient client = HttpClient.newHttpClient();

    public String getEndpoint() {
        return Config.TARGET_IS_ALIVE_ROUTE;
    }

    public boolean check() throws IOException {
        if (Config.TARGET_IS_ALIVE_ROUTE != null) {
            try {
                final var request = HttpRequest
                    .newBuilder()
                    .version(HttpClient.Version.HTTP_1_1)
                    .GET()
                    .uri(URI.create(Config.TARGET_IS_ALIVE_ROUTE))
                    .build();

                var targetIsAliveResponse = client.send(request, HttpResponse.BodyHandlers.ofString());
                if (targetIsAliveResponse.statusCode() != 200) {
                    Monitor.targetNotAlive(targetIsAliveResponse.statusCode());
                    return false;
                }
                Monitor.targetAlive(targetIsAliveResponse.statusCode());
                return true;
            } catch (Exception e) {
                Monitor.initializationError(e);
                return false;
            }
        }
        return true;
    }
}
