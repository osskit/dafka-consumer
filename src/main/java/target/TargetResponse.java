package target;

import java.util.OptionalLong;

public class TargetResponse {

    public OptionalLong callLatency;
    public OptionalLong resultLatency;

    public TargetResponse() {}

    TargetResponse(OptionalLong callLatency, OptionalLong resultLatency) {
        this.callLatency = callLatency;
        this.resultLatency = resultLatency;
    }
}
