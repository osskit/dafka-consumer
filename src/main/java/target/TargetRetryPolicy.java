package target;

import configuration.Config;
import dev.failsafe.RetryPolicy;
import dev.failsafe.event.ExecutionAttemptedEvent;
import dev.failsafe.event.ExecutionCompletedEvent;
import dev.failsafe.okhttp.FailsafeCall;
import java.io.IOException;
import java.net.ConnectException;
import java.net.SocketException;
import java.net.UnknownHostException;
import java.time.Duration;
import java.time.temporal.ChronoUnit;
import java.util.Optional;
import monitoring.Monitor;
import okhttp3.Response;
import org.apache.kafka.clients.consumer.ConsumerRecord;

public class TargetRetryPolicy {

    public static FailsafeCall.FailsafeCallBuilder create(ConsumerRecord<String, String> record, String requestId) {
        var delay = Config.RETRY_POLICY_EXPONENTIAL_BACKOFF.get(0);
        var maxDelay = Config.RETRY_POLICY_EXPONENTIAL_BACKOFF.get(1);
        var delayFactor = Config.RETRY_POLICY_EXPONENTIAL_BACKOFF.get(2);
        var maxDuration = Duration.ofMillis(Config.RETRY_POLICY_MAX_DURATION_MS);

        Monitor.processMessageStarted(record, requestId);

        var connectionFailureDelay = Config.CONNECTION_FAILURE_RETRY_POLICY_EXPONENTIAL_BACKOFF.get(0);
        var connectionFailureMaxDelay = Config.CONNECTION_FAILURE_RETRY_POLICY_EXPONENTIAL_BACKOFF.get(1);
        var connectionFailureDelayFactor = Config.CONNECTION_FAILURE_RETRY_POLICY_EXPONENTIAL_BACKOFF.get(2);
        var connectionFailureMaxDuration = Duration.ofMillis(Config.CONNECTION_FAILURE_RETRY_POLICY_MAX_DURATION_MS);

        var connectionRetryPolicy = RetryPolicy
            .<Response>builder()
            .withBackoff(
                connectionFailureDelay,
                connectionFailureMaxDelay,
                ChronoUnit.MILLIS,
                connectionFailureDelayFactor
            )
            .withMaxDuration(connectionFailureMaxDuration)
            .withMaxAttempts(Config.CONNECTION_FAILURE_RETRY_POLICY_MAX_RETRIES)
            .handle(IOException.class)
            .handleResultIf(r -> Integer.toString(r.code()).matches("503"))
            .onRetry(e -> {
                Monitor.targetConnectionRetry(
                    extractAttemptedResponseBody(e),
                    e.getLastException(),
                    e.getAttemptCount(),
                    requestId
                );
            })
            .build();

        var executionRetryPolicy = dev.failsafe.RetryPolicy
            .<Response>builder()
            .withBackoff(delay, maxDelay, ChronoUnit.MILLIS, delayFactor)
            .withMaxDuration(maxDuration)
            .withMaxAttempts(Config.RETRY_POLICY_MAX_RETRIES)
            .handleIf(e -> false)
            .handleResultIf(r -> Integer.toString(r.code()).matches(Config.RETRY_PROCESS_WHEN_STATUS_CODE_MATCH))
            .onRetry(e -> {
                Monitor.targetExecutionRetry(
                    extractAttemptedResponseBody(e),
                    e.getLastException(),
                    e.getAttemptCount(),
                    requestId
                );
            })
            .build();

        return FailsafeCall.with(executionRetryPolicy, connectionRetryPolicy);
    }

    private static Optional<String> extractAttemptedResponseBody(ExecutionAttemptedEvent<Response> e) {
        return Optional
            .ofNullable(e.getLastResult())
            .flatMap(r -> {
                try {
                    try (Response response = r) {
                        return Optional.of(response.body().string());
                    }
                } catch (IOException error) {
                    return Optional.empty();
                }
            });
    }
}
