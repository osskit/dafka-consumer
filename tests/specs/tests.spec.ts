import type {Orchestrator} from '../testcontainers/orchestrator.js';
import {start} from '../testcontainers/orchestrator.js';
import {getCalls, mockHttpTarget, mockFaultyHttpTarget} from '../services/target.js';
import {getOffset} from '../services/getOffset.js';
import {produce} from '../services/produce.js';
import {range} from 'lodash-es';
import pRetry from 'p-retry';
import delay from 'delay';

const topicRoutes = (topicRoutes: {topic: string; targetPath: string}[]) =>
    topicRoutes.map(({topic, targetPath}) => `${topic}:${targetPath}`).join(',');

describe('tests', () => {
    let orchestrator: Orchestrator;

    beforeEach(async () => {
        orchestrator = await start();
    }, 5 * 60 * 1000);

    afterEach(async () => {
        orchestrator.stop();
    });

    describe('basic', () => {
        it('should produce and consume', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: 'foo', targetPath: '/consume'}]),
                },
                ['foo']
            );
            const target = await mockHttpTarget(orchestrator.target, '/consume', 200);

            //act
            await produce(orchestrator, {
                topic: 'foo',
                messages: [{value: JSON.stringify({data: 'foo'})}],
            });

            //assert
            await expect(getCalls(orchestrator.target, target)).resolves.toMatchSnapshot();
            await expect(getOffset(orchestrator.admin(), 'foo')).resolves.toBe(1);
        });

        it('should produce and consume with regex patterns', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: '^([^.]+).foo', targetPath: '/consume'}]),
                },
                ['prefix.foo']
            );
            const target = await mockHttpTarget(orchestrator.target, '/consume', 200);

            //act
            await produce(orchestrator, {
                topic: 'prefix.foo',
                messages: [{value: JSON.stringify({data: 'foo'})}],
            });

            //assert
            await expect(getCalls(orchestrator.target, target)).resolves.toMatchSnapshot();
            await expect(getOffset(orchestrator.admin(), 'prefix.foo')).resolves.toBe(1);
        });

        it.only('should consume bursts of records', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: 'foo', targetPath: '/consume'}]),
                },
                ['foo']
            );
            await mockHttpTarget(orchestrator.target, '/consume', 200);

            //act
            await produce(orchestrator, {
                topic: 'foo',
                messages: range(10000).map((_) => ({value: JSON.stringify({data: 'foo'})})),
            });
            await delay(30000);

            //assert
            await expect(getOffset(orchestrator.admin(), 'foo')).resolves.toBe(10000);
        });
    });

    describe('protocol', () => {
        it('should add record headers to target call', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: 'foo', targetPath: '/consume'}]),
                },
                ['foo']
            );
            const target = await mockHttpTarget(orchestrator.target, '/consume', 200);

            //act
            await produce(orchestrator, {
                topic: 'foo',
                messages: [
                    {
                        value: JSON.stringify({data: 'foo'}),
                        headers: {
                            'x-request-id': '123',
                            'x-b3-traceid': '456',
                            'x-b3-spanid': '789',
                            'x-b3-parentspanid': '101112',
                            'x-b3-sampled': '1',
                            'x-b3-flags': '1',
                            'x-ot-span-context': 'foo',
                        },
                    },
                ],
            });

            //assert
            await expect(getCalls(orchestrator.target, target, true)).resolves.toMatchSnapshot();
        });

        it('copy body headers to record headers', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: 'foo', targetPath: '/consume'}]),
                    BODY_HEADERS_PATHS: 'bla,baz',
                },
                ['foo']
            );
            const target = await mockHttpTarget(orchestrator.target, '/consume', 200);

            //act
            await produce(orchestrator, {
                topic: 'foo',
                messages: [
                    {
                        value: JSON.stringify({data: 'foo', bla: {foo: 'bar', baz: null}}),
                    },
                ],
            });

            //assert
            await expect(getCalls(orchestrator.target, target, true)).resolves.toMatchSnapshot();
        });
    });

    describe('resiliency', () => {
        it('consumer should wait after connection error from target', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: 'foo', targetPath: '/consume'}]),
                    CONNECTION_FAILURE_RETRY_POLICY_EXPONENTIAL_BACKOFF: '5,8000000,2',
                    CONNECTION_FAILURE_RETRY_POLICY_MAX_DURATION_MS: '8000000',
                },
                ['foo']
            );
            const target = await mockFaultyHttpTarget(orchestrator.target, '/consume');

            //act
            await produce(orchestrator, {
                topic: 'foo',
                messages: [{value: JSON.stringify({data: 'foo'})}],
            });
            await delay(20000);

            //assert
            await expect(getCalls(orchestrator.target, target)).resolves.toHaveLength(10);
            await expect(getOffset(orchestrator.admin(), 'foo')).resolves.toBe(1);
        });

        it('consumer should wait after getting 503 response from target', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: 'foo', targetPath: '/consume'}]),
                    CONNECTION_FAILURE_RETRY_POLICY_EXPONENTIAL_BACKOFF: '50,5000,2',
                    CONNECTION_FAILURE_RETRY_POLICY_MAX_DURATION_MS: '5000',
                },
                ['foo']
            );
            const target = await mockHttpTarget(orchestrator.target, '/consume', 503);

            //act
            await produce(orchestrator, {
                topic: 'foo',
                messages: [{value: JSON.stringify({data: 'foo'})}],
            });
            await delay(20000);

            //assert
            await expect(getCalls(orchestrator.target, target)).resolves.toHaveLength(10);
            await expect(getOffset(orchestrator.admin(), 'foo')).resolves.toBe(1);
        });

        it('consumer should recover from socket error', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: 'foo', targetPath: '/consume'}]),
                    CONNECTION_FAILURE_RETRY_POLICY_EXPONENTIAL_BACKOFF: '50,5000,2',
                    CONNECTION_FAILURE_RETRY_POLICY_MAX_DURATION_MS: '5000',
                },
                ['foo']
            );
            await mockFaultyHttpTarget(orchestrator.target, '/consume');

            //act
            await produce(orchestrator, {
                topic: 'foo',
                messages: [{value: JSON.stringify({data: 'foo'})}],
            });
            await delay(2000);
            await orchestrator.target.reset();
            const target = await mockHttpTarget(orchestrator.target, '/consume', 200);
            await delay(2000);

            //assert
            await expect(getCalls(orchestrator.target, target)).resolves.toHaveLength(1);
            await expect(getOffset(orchestrator.admin(), 'foo')).resolves.toBe(1);
        });

        it('consumer should produce to retry topic', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: 'foo', targetPath: '/consume'}]),
                    RETRY_TOPIC: 'retry',
                    RETRY_PROCESS_WHEN_STATUS_CODE_MATCH: '511',
                    PRODUCE_TO_RETRY_TOPIC_WHEN_STATUS_CODE_MATCH: '511',
                    RETRY_POLICY_EXPONENTIAL_BACKOFF: '50,500,10',
                    RETRY_POLICY_MAX_DURATION: '1000',
                },
                ['foo', 'retry']
            );
            const target = await mockHttpTarget(orchestrator.target, '/consume', 511);

            //act
            await produce(orchestrator, {
                topic: 'foo',
                messages: [{value: JSON.stringify({data: 'foo'})}],
            });
            await delay(5000);

            //assert
            await expect(getCalls(orchestrator.target, target)).resolves.toHaveLength(10);
            await expect(getOffset(orchestrator.admin(), 'foo')).resolves.toBe(1);
            await expect(orchestrator.consume('retry')).resolves.toMatchSnapshot();
        });

        it('consumer should produce to dead letter topic', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: 'foo', targetPath: '/consume'}]),
                    DEAD_LETTER_TOPIC: 'dead',
                    PRODUCE_TO_DEAD_LETTER_TOPIC_WHEN_STATUS_CODE_MATCH: '428',
                },
                ['foo', 'dead']
            );
            const target = await mockHttpTarget(orchestrator.target, '/consume', 428);

            //act
            await produce(orchestrator, {
                topic: 'foo',
                messages: [{value: JSON.stringify({data: 'foo'})}],
            });
            await delay(5000);

            //assert
            await expect(getCalls(orchestrator.target, target)).resolves.toHaveLength(10);
            await expect(getOffset(orchestrator.admin(), 'foo')).resolves.toBe(1);
            await expect(orchestrator.consume('dead')).resolves.toMatchSnapshot();
        });

        it('consumer should produce to dead letter topic when value is not valid JSON', async () => {
            //prepare
            await orchestrator.consumer(
                {
                    GROUP_ID: 'test',
                    TARGET_BASE_URL: 'http://mocks:8080',
                    TOPICS_ROUTES: topicRoutes([{topic: 'foo', targetPath: '/consume'}]),
                    DEAD_LETTER_TOPIC: 'dead',
                },
                ['foo', 'dead']
            );

            //act
            await produce(orchestrator, {
                topic: 'foo',
                messages: [{value: 'wat'}],
            });
            await delay(5000);

            //assert
            // await expect(getOffset(orchestrator.admin(), 'foo')).resolves.toBe(-1);
            await expect(orchestrator.consume('dead')).resolves.toMatchSnapshot();
        });
    });
});
