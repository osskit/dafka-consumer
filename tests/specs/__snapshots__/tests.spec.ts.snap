// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests should add record headers to target call 1`] = `
Object {
  "body": Object {
    "data": "foo",
  },
  "headers": Object {
    "content-length": "14",
    "content-type": "application/json",
    "host": "target:2000",
    "user-agent": "Java-http-client/13-ea",
    "x-b3-flags": "1",
    "x-b3-parentspanid": "101112",
    "x-b3-sampled": "1",
    "x-b3-spanid": "789",
    "x-b3-traceid": "456",
    "x-ot-span-context": "foo",
    "x-record-offset": Any<String>,
    "x-record-original-topic": "foo",
    "x-record-partition": "3",
    "x-record-timestamp": Any<String>,
    "x-record-topic": "foo",
    "x-request-id": "123",
  },
  "method": "POST",
  "path": "/consume",
}
`;

exports[`tests should produce and consume 1`] = `
Object {
  "body": Object {
    "data": "foo",
  },
  "headers": Object {
    "content-length": "14",
    "content-type": "application/json",
    "host": "target:2000",
    "user-agent": "Java-http-client/13-ea",
    "x-record-offset": Any<String>,
    "x-record-original-topic": "foo",
    "x-record-partition": "3",
    "x-record-timestamp": Any<String>,
    "x-record-topic": "foo",
  },
  "method": "POST",
  "path": "/consume",
}
`;

exports[`tests should transform and add cloud event headers to target call 1`] = `
Object {
  "body": Object {
    "data": "foo",
  },
  "headers": Object {
    "ce-id": "id",
    "ce-source": "source",
    "ce-specversion": "1",
    "ce-time": "123456",
    "ce-type": "type",
    "content-length": "14",
    "content-type": "application/json",
    "host": "target:2000",
    "random_header": "random",
    "user-agent": "Java-http-client/13-ea",
    "x-record-offset": Any<String>,
    "x-record-original-topic": "foo",
    "x-record-partition": "3",
    "x-record-timestamp": Any<String>,
    "x-record-topic": "foo",
    "x-request-id": "bla",
  },
  "method": "POST",
  "path": "/consume",
}
`;
